File: fragments/api/fastify/blockchain/solana_round.ts
import { type FastifyInstance, type FastifyPluginOptions } from "fastify";
import { Address, address, generateKeyPairSigner, type KeyPairSigner } from "@solana/kit";
import {
  activateRound,
  completeRound,
  getRoundAccount,
  initialiseRound,
} from "../../../solana_program_round/solana_round_interface";
import { getEnvVar } from "../../../env_vars/env_vars_utils";
import { sendAndConfirmAirdrop } from "../../../solana_airdrop/solana_airdrop_utils";
import { confirmRecentSignature } from "../../../solana_transaction/solana_transaction_utils";
import { initRpcClient } from "../../../solana_rpc/solana_rpc_utils";
import { LAMPORTS_PER_SOL } from "@solana/web3.js";
import { isSome, unwrapOption } from "@solana/kit";

// In-memory storage for keypairs
// In production, use a secure key management service or encrypted database
const keypairStorage: Record<string, KeyPairSigner> = {};

export const routes = (fastify: FastifyInstance, _: FastifyPluginOptions) => {
  const client = initRpcClient();

  fastify.post<{
    Reply: {
      200: { address: string; start_slot: string };
      400: { error: string };
      500: { error: string };
    };
  }>("/round/initialise", async (request, reply) => {
    try {
      const programAddress = getProgramAddress();

      const signer = await generateKeyPairSigner();
      await sendAndConfirmAirdrop(signer.address, BigInt(LAMPORTS_PER_SOL));
      keypairStorage[signer.address] = signer;
      request.log.info({ address: signer.address }, "Keypair stored for round authority");

      const recentSlot = await client.getSlot({ commitment: "confirmed" }).send();
      const startSlot = recentSlot + 3n;

      const signature = await initialiseRound(signer, programAddress, startSlot);
      const confirmed = await confirmRecentSignature(signature);

      if (!confirmed) {
        return reply.code(500).send({ error: "Transaction sent but confirmation timed out" });
      }

      return reply.code(200).send({
        address: signer.address,
        start_slot: startSlot.toString(),
      });
    } catch (error) {
      request.log.error(error, "Error initialising round");
      return reply.code(500).send({ error: "Internal Server Error" });
    }
  });

  fastify.get<{
    Params: { address: string };
    Reply: {
      200: {
        start_slot: string;
        authority: string;
        activated_at: string | null;
        activated_by: string | null;
        completed_at: string | null;
      };
      404: void;
      500: { error: string };
    };
  }>("/round/:address", async (request, reply) => {
    try {
      const { address } = request.params;
      const programAddress = getProgramAddress();
      const keypair = keypairStorage[address];

      if (!keypair) {
        return reply.code(404).send();
      }

      const account = await getRoundAccount(keypair.address, programAddress);

      return reply.code(200).send({
        start_slot: account.start_slot.toString(),
        authority: account.authority,
        activated_at: isSome(account.activated_at) && unwrapOption(account.activated_at)?.toString() || null,
        activated_by: isSome(account.activated_by) && unwrapOption(account.activated_by) || null,
        completed_at: isSome(account.completed_at) && unwrapOption(account.completed_at)?.toString() || null,
      });
    } catch (error) {
      request.log.error(error, "Error fetching round account");
      return reply.code(500).send({ error: "Internal Server Error" });
    }
  });

  fastify.patch<{
    Params: { address: string };
    Reply: {
      200: void;
      404: void;
      500: { error: string };
    };
  }>("/round/:address/activate", async (request, reply) => {
    try {
      const { address: roundAddress } = request.params;
      const programAddress = getProgramAddress();
      const roundExists = keypairStorage[roundAddress];

      if (!roundExists) {
        return reply.code(404).send();
      }

      const signer = await generateKeyPairSigner();
      await sendAndConfirmAirdrop(signer.address, BigInt(LAMPORTS_PER_SOL));

      const signature = await activateRound(signer, programAddress, address(roundAddress));
      const confirmed = await confirmRecentSignature(signature);

      if (!confirmed) {
        return reply.code(500).send({ error: "Transaction sent but confirmation timed out" });
      }

      return reply.code(200).send();
    } catch (error) {
      request.log.error(error, "Error activating round");
      return reply.code(500).send({ error: "Internal Server Error" });
    }
  });

  fastify.patch<{
    Params: { address: string };
    Reply: {
      200: void;
      404: void;
      500: { error: string };
    };
  }>("/round/:address/complete", async (request, reply) => {
    try {
      const { address } = request.params;
      const programAddress = getProgramAddress();
      const keypair = keypairStorage[address];

      if (!keypair) {
        return reply.code(404).send();
      }

      const signature = await completeRound(keypair, programAddress);
      const confirmed = await confirmRecentSignature(signature);

      if (!confirmed) {
        return reply.code(500).send({ error: "Transaction sent but confirmation timed out" });
      }

      return reply.code(200).send();
    } catch (error) {
      request.log.error(error, "Error completing round");
      return reply.code(500).send({ error: "Internal Server Error" });
    }
  });
};

const getProgramAddress = (): Address => {
  const programId = getEnvVar("round_PROGRAM_ID");
  if (!programId) {
    throw new Error("environment variable round_PROGRAM_ID is not set");
  }
  return address(programId);
};

File: fragments/apis/fastapi/blockchain/solana_round.py
import logging
from fastapi import APIRouter, HTTPException
from solders.keypair import Keypair
from solders.pubkey import Pubkey
from solana.constants import LAMPORTS_PER_SOL

from fragments.solana_program_round import initialise_round, get_round_account, activate_round, complete_round
from fragments.solana_airdrop import send_and_confirm_airdrop
from fragments.solana_transaction import confirm_recent_signature
from fragments.solana_rpc import init_rpc_client
from fragments.env_vars import get_env_var

solana_round_router = APIRouter()
logger = logging.getLogger(__name__)

# In-memory storage for keypairs
# In production, use a secure key management service or encrypted database
keypair_storage: dict[str, Keypair] = {}


def get_program_address() -> Pubkey:
    program_id = get_env_var("round_PROGRAM_ID")
    if program_id is None:
        raise ValueError("environment variable round_PROGRAM_ID is not set")
    return Pubkey.from_string(program_id)


@solana_round_router.post("/initialise", status_code=200)
async def initialise_round_route():
    try:
        program_address = get_program_address()
        signer = Keypair()
        address = str(signer.pubkey())
        await send_and_confirm_airdrop(signer.pubkey(), LAMPORTS_PER_SOL)
        keypair_storage[address] = signer
        logger.info("Keypair stored for round authority: %s", address)

        client = init_rpc_client()
        recent_slot = (await client.get_slot()).value
        start_slot = recent_slot + 3

        signature = await initialise_round(signer, program_address, start_slot)
        confirmed = await confirm_recent_signature(signature)

        if not confirmed:
            raise HTTPException(
                status_code=500,
                detail="Transaction sent but confirmation timed out",
            )

        return {"address": address, "start_slot": str(start_slot)}
    except HTTPException:
        raise
    except Exception as e:
        logger.error("Error initialising round: %s", e)
        raise HTTPException(status_code=500, detail="Internal Server Error") from e


@solana_round_router.get("/{address}", status_code=200)
async def get_round(address: str):
    try:
        keypair = keypair_storage.get(address)
        if keypair is None:
            raise HTTPException(status_code=404)

        program_address = get_program_address()
        account = await get_round_account(keypair.pubkey(), program_address)

        return {
            "start_slot": str(account["start_slot"]),
            "authority": str(account["authority"]),
            "activated_at": str(account["activated_at"]) if account["activated_at"] is not None else None,
            "activated_by": str(account["activated_by"]) if account["activated_by"] is not None else None,
            "completed_at": str(account["completed_at"]) if account["completed_at"] is not None else None,
        }
    except HTTPException:
        raise
    except Exception as e:
        logger.error("Error fetching round account: %s", e)
        raise HTTPException(status_code=500, detail="Internal Server Error") from e


@solana_round_router.patch("/{address}/activate", status_code=200)
async def activate_round_route(address: str):
    try:
        round_authority = keypair_storage.get(address)
        if round_authority is None:
            raise HTTPException(status_code=404)

        program_address = get_program_address()

        payer = Keypair()
        await send_and_confirm_airdrop(payer.pubkey(), LAMPORTS_PER_SOL)

        signature = await activate_round(payer, program_address, round_authority.pubkey())
        confirmed = await confirm_recent_signature(signature)

        if not confirmed:
            raise HTTPException(
                status_code=500,
                detail="Transaction sent but confirmation timed out",
            )

        return None
    except HTTPException:
        raise
    except Exception as e:
        logger.error("Error activating round: %s", e)
        raise HTTPException(status_code=500, detail="Internal Server Error") from e


@solana_round_router.patch("/{address}/complete", status_code=200)
async def complete_round_route(address: str):
    try:
        keypair = keypair_storage.get(address)
        if keypair is None:
            raise HTTPException(status_code=404)

        program_address = get_program_address()

        signature = await complete_round(keypair, program_address)
        confirmed = await confirm_recent_signature(signature)

        if not confirmed:
            raise HTTPException(
                status_code=500,
                detail="Transaction sent but confirmation timed out",
            )

        return None
    except HTTPException:
        raise
    except Exception as e:
        logger.error("Error completing round: %s", e)
        raise HTTPException(status_code=500, detail="Internal Server Error") from e
